### DISCUSSION TITLE
Use a pipette cursor when color picking

### DISCUSSION THREAD

**Discussion by bsil78** [godotengine/godot-proposals#10050] [Editor]
Hello everyone !

I am currently using Godot 4.3 in beta state, and I found out that color picker has lost its cursor modification when color picking ; it bothered me somewhat as I had no feedback to know if I was in color picking mode.

In Godot 3, it was useful to see cursor becoming a hand with index figer pointing up, showing color picking is active.

What could be nice it to get back this pointer modification when color picking ; more over it should be a dedicated pointer like a pipette as we see in other software.

IMHO, that would be a nice enhancement.

What do you think ?

(PS : I attached a pipette.svg file I made with inkscape for somebody to use in implementation eventually)

![pipette](https://github.com/godotengine/godot-proposals/assets/13476087/9c4aa0d2-130b-4f15-9ff8-a8ca835688ae)

**Comment by Calinou** [comment]
There is no built-in pipette cursor in operating systems' themes, so we'd need to design a custom cursor which may look inconsistent with the defaults on some OSes (on top of [suffering from scaling issues](https://github.com/godotengine/godot-proposals/issues/9262)). This is particularly the case on Windows where users can choose any cursor color they wish now (check the Accessibility options).

Personally, I'd prefer using the crosshair during color picking, which *is* a standard OS cursor.

Also, I don't think Godot can currently override the mouse cursor when the mouse cursor is *not* hovering a Godot window. This would be required for consistent appearance regardless of whether you're hovering a Godot window, as Godot 4 can pick colors located outside its window.

**Comment by bsil78** [comment]
Just having another cursor is enough IMHO. Crosshair is a good alternative ! 👍

**Comment by Mickeon** [comment]
> In Godot 3, it was useful to see cursor becoming a hand with index figer pointing up, showing color picking is active.

It turns out that bits of code to change the cursor's shape _still_ exist but for the reason above they no longer work. With them not being removed or explained in comments, the current behavior _may_ be an oversight.

https://github.com/godotengine/godot/pull/93803

**Comment by bsil78** [comment]
As [your PR](https://github.com/godotengine/godot/pull/93803) mention, picking color outside of Godot is possible which I did not even noticed before ! the PR is a first step good enough for now, but I wonder why cursor form is not kept when outside. It should, else we cannot know picking outside is possible...

**Reply by Mickeon** [reply]
> But I wonder why cursor firm is not kept when outside.

I don't believe any proper application has control of it outside of their windows. If they did, it would be particularly error-prone.
I can imagine one very "stupid" way to accomplish this goal would be by covering all screens with an invisible, but still tangible Godot window, so long as color-picking is enabled.

> Else we cannot know picking outside is possible...

The tooltip hints it, but that's really not enough.

**Reply by bsil78** [reply]
Oh yes, I see the problem. Ok, I think the solution you are proposing is the good one, just instantiate a transparent exclusive popup window and close it on a clic, thus all screen would get the picking color cursor 👍

**Reply by KoBeWi** [reply]
It works in GIMP:

https://github.com/godotengine/godot-proposals/assets/2223172/c6d25402-a645-4f55-b423-cbb2563cebac

So it's possible.

**Comment by bsil78** [comment]
@Calinou  : if we use a popup exlusive window that covers all the screen, we could use a pipette cursor as a sprite (hiding OS cursor while operation) ; that way any OS would get same dedicated nice pipette cursor ? isn't it possible ?

**Reply by Mickeon** [reply]
The problems with a custom cursor that Calinou described above would still apply. More importantly to me, nothing in the editor ever uses a custom cursor, so it would be a major outlier.

**Reply by bsil78** [reply]
How would it apply if it is a sprite, not the actual OS cursor ?
I thought about the "outlier", but why a major outlier ? A custom cursor for a specific need is not so bad, isn't it ? I am not sure about that.
